### PR TITLE
rendering on jupyter xeus-cling kernel

### PR DESCRIPTION
**Pull request by isuruf**
Fixes https://github.com/symengine/symengine/issues/1510

### CODE REVIEW

**Comment by isuruf** [comment]
Here's the binder link for the new notebook https://mybinder.org/v2/gh/isuruf/symengine/render?filepath=notebooks%2Fsymengine.ipynb

File: symengine/expression.h [hunk 1]
```diff
@@ -305,7 +312,22 @@ struct print_coefficient_impl<U, typename std::
...
 };
-}
+} // namespace piranha
 #endif // HAVE_SYMENGINE_PIRANHA
 
+// Utility functions for xeus-cling
+namespace SymEngine
+{
+
+#include <nlohmann/json.hpp>
```
Comment (certik) [inline symengine/expression.h]: Do we need to hook this up into CMake to check that JSON is installed?

Comment (isuruf) [reply symengine/expression.h]: Fixed with a `#ifdef`

**Comment by certik** [comment]
> Here's the binder link for the new notebook https://mybinder.org/v2/gh/isuruf/symengine/render?filepath=notebooks%2Fsymengine.ipynb

The link fails for me with:

screenshot_2019-02-11 binder beta

(I can't copy the error text, or see the full error...).

File: notebooks/symengine.ipynb [hunk 2]
```diff
@@ -33,21 +32,77 @@
...
+      ],
+      "text/plain": [
+       "(x + sqrt(2))**10"
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
```
Comment (certik) [inline notebooks/symengine.ipynb]: I would keep the notebook unevaluated in git.

**Comment by isuruf** [comment]
@certik, can you check the link https://mybinder.org/v2/gh/isuruf/symengine/render?filepath=notebooks%2Fsymengine.ipynb again?

**Comment by certik** [comment]
@isuruf the binder takes forever to show anything. I wonder if we should somehow pre-build a docker image to make that faster to load.

Finally got it to load. I think it works for me:

https://nbviewer.jupyter.org/gist/certik/448a7b18e567c2036068720feca00bcb

**Comment by SylvainCorlay** [comment]
Wait what!!!

**Comment by SylvainCorlay** [comment]
giphy

**Comment by SylvainCorlay** [comment]
> I wonder if we should somehow pre-build a docker image to make that faster to load.

Binder does just that. I takes much longer someone connects to the URL.

Also, the binder image takes less time to load if it was loaded recently.

**Comment by certik** [comment]
> > I wonder if we should somehow pre-build a docker image to make that faster to load.
> 
> Binder does just that. I takes much longer someone connects to the URL.
> 
> Also, the binder image takes less time to load if it was loaded recently.

I think binder builds it only when somebody clicks on the link, and maybe more often. And in symengine's case, I frequently have to wait for something like 5 minutes.

What I have in mind is that we prebuilt the docker image ahead of time, upload online, and binder just pulls it, which will make it a lot faster.

**Comment by isuruf** [comment]
Binder builds it once for each commit and saves the image. I like this way, because I can check an arbitrary commit in any fork. Otherwise only master of symengine/symengine can be checked.

**Comment by SylvainCorlay** [comment]
For e.g. the xeus-cling repositories we have the following workflow:

- The release tag of xeus-cling includes the change to environment.yml to pull the tagged release.
- The binder links points to the `stable` branch, which we set to that tag when the conda packages are built.

https://mybinder.org/v2/gh/QuantStack/xeus-cling/stable?filepath=notebooks/xcpp.ipynb

**Comment by SylvainCorlay** [comment]
The PR looks good to me. The test failure is a network error and the binder works.

**Comment by SylvainCorlay** [comment]
I am super excited about this and would love to integrate this into the xeus-cling binder. Do you guys plan on making a release with this when it gets in?

**Comment by isuruf** [comment]
@SylvainCorlay, yes, we'll do a release after #1504

**Pull request merged**